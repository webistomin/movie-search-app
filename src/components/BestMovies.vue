<template>
  <section class="best">
    <div class="best__block">
      <h2 class="best__title title">Today best choice</h2>
    </div>
    <ul class="best__list">
      <li class="best__item">
        <a href="" class="best__link">
          <picture class="best__pic best__pic--gold">
            <img src="img/best@1x.jpg" alt="#" width="41" height="41" class="best__img">
          </picture>
          <div class="best__block best__block--padding">
            <span class="best__name">Gravity</span>
            <div class="best__rating"></div>
          </div>
        </a>
      </li>
      <li class="best__item">
        <a href="#" class="best__link">
          <picture class="best__pic best__pic--gold">
            <img src="img/best@1x.jpg" alt="#" width="41" height="41" class="best__img">
          </picture>
          <div class="best__block best__block--padding">
            <span class="best__name">Gravity</span>
            <div class="best__rating"></div>
          </div>
        </a>
      </li>
      <li class="best__item">
        <a href="#" class="best__link">
          <picture class="best__pic best__pic--gold">
            <img src="img/best@1x.jpg" alt="#" width="41" height="41" class="best__img">
          </picture>
          <div class="best__block best__block--padding">
            <span class="best__name">Gravity</span>
            <div class="best__rating"></div>
          </div>
        </a>
      </li>
      <li class="best__item">
        <a href="#" class="best__link">
          <picture class="best__pic best__pic--silver">
            <img src="img/best@1x.jpg" alt="#" width="41" height="41" class="best__img">
          </picture>
          <div class="best__block best__block--padding">
            <span class="best__name">Gravity</span>
            <div class="best__rating"></div>
          </div>
        </a>
      </li>
      <li class="best__item">
        <a href="#" class="best__link">
          <picture class="best__pic best__pic--silver">
            <img src="img/best@1x.jpg" alt="#" width="41" height="41" class="best__img">
          </picture>
          <div class="best__block best__block--padding">
            <span class="best__name">Gravity</span>
            <div class="best__rating"></div>
          </div>
        </a>
      </li>
    </ul>
  </section>
</template>

<script>
  export default {
    name: 'BestMovies',
  };
</script>

<style lang="sass">
  .best
    display: none
    background-color: $color-brown--secondary

    &__title
      padding-top: 24px
      padding-bottom: 24px
      text-align: center
      background-color: $color-brown

    &__list
      display: flex
      flex-direction: column
      align-items: center
      padding-top: 20px

    &__item
      display: flex
      min-width: 122px
      margin-bottom: 20px

    &__link
      display: flex

      &:hover,
      &:focus
        .best__name
          color: $color-text--secondary
          transition: all 0.3s ease-in-out

      &:active
        .best__name
          color: $color-action
          transition: all 0.3s ease-in-out

    &__name
      display: block
      max-width: 63px
      overflow: hidden
      text-overflow: ellipsis
      font-size: 12px
      color: $color-text
      margin-bottom: 8px
      transition: all 0.3s ease-in-out

    &__img
      border-radius: 4px

    &__pic
      position: relative
      margin-right: 10px

      &::before
        width: 3px
        height: 100%
        content: ""
        position: absolute
        left: 0
        top: 0

      &--gold
        &::before
          background-color: $color-gold

      &--silver
        &::before
          background-color: $color-silver

    &__rating
      width: 63px
      height: 9px
      background-color: transparent
      background-image: url("~/static/img/content/stars.svg")
      background-size: contain
      background-repeat: no-repeat

    &__block
      &--padding
        padding: 5px 0

    @media (min-width: $screen-xs)
      &__list
        display: flex
        flex-direction: row
        flex-wrap: wrap
        justify-content: space-around

    @media (min-width: $screen-sm)
      display: block

      &__list
        padding-top: 25px

      &__item
        margin-bottom: 25px

    @media (min-width: $screen-md)
      &__title
        text-align: left
        padding-top: 23px
        padding-bottom: 22px
        padding-left: 44px
</style>
