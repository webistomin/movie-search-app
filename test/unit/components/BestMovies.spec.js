import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import BestMovies from '../../../src/components/BestMovies';
import routes from '../../../src/router/';

describe('BestMovies.vue', () => {
  let actions;
  let store;
  let state;
  let mutations;
  let getters;
  let router;
  let vm;
  let mockedRoute;
  let localVue;

  beforeEach(() => {
    localStorage.clear();
    jest.clearAllMocks();
    jest.resetModules();

    actions = {
      fetchPopularMovies: jest.fn().mockImplementation(() => Promise.resolve('called')),
    };

    mutations = {
      setLoadingState: (state, payload) => {
        state.loadingState = payload;
      },
    };

    state = {
      loadingState: null,
      popularMovies: [{ id: 0, title: 'Prison Break' }],
    };

    getters = {
      getPopularMovies: state => state.popularMovies,
    };

    store = new Vuex.Store({
      state,
      mutations,
      actions,
      getters,
    });

    localVue = createLocalVue();
    localVue.use(Vuex);
    router = routes;
  });

  it('Matches snapshot', () => {
    vm = shallowMount(BestMovies, {
      store,
      localVue,
      router,
    }).vm;
    expect(vm.$el).toMatchSnapshot();
  });

  describe('test "fetchPopularMovies" action', () => {
    it('if route.name !== "popular", then call action', () => {
      vm = shallowMount(BestMovies, {
        store,
        localVue,
        router,
      }).vm;
      expect(actions.fetchPopularMovies).toHaveBeenCalled();
    });

    it('if route.name === "popular", then doesnt call action', () => {
      mockedRoute = { path: '/popular', name: 'popular' };
      vm = shallowMount(BestMovies, {
        store,
        localVue,
        router,
        beforeCreate() {
          this._route = mockedRoute;
        },
      }).vm;
      expect(actions.fetchPopularMovies).not.toHaveBeenCalled();
    });

    it('if route.name !== "popular" and !== "movie", then commit mutation', async () => {
      mockedRoute = { path: '/', name: 'Home' };
      vm = shallowMount(BestMovies, {
        store,
        localVue,
        router,
        beforeCreate() {
          this._route = mockedRoute;
        },
      }).vm;
      await actions.fetchPopularMovies;
      expect(state.loadingState).toEqual(false);
    });

    it('if route.name === "movie", then doesnt commit mutation', async () => {
      mockedRoute = { path: '/movie/297802', name: 'movie' };
      vm = shallowMount(BestMovies, {
        store,
        localVue,
        router,
        beforeCreate() {
          this._route = mockedRoute;
        },
      }).vm;
      await actions.fetchPopularMovies;
      expect(state.loadingState).toEqual(null);
    });

    it('getPopularMovies getter work correctly', () => {
      vm = shallowMount(BestMovies, {
        store,
        localVue,
        router,
      }).vm;
      expect(vm.getPopularMovies).toEqual([{ id: 0, title: 'Prison Break' }]);
    });
  });
});
