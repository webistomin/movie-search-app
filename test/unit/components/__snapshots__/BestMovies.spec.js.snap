// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BestMovies.vue Matches snapshot 1`] = `
<section
  class="best"
>
  <div
    class="best__block"
  >
    <h2
      class="best__title title"
    >
      Today best choice
    </h2>
  </div>
   
  <bestlist-stub
    movieslist="[object Object]"
  />
</section>
`;
